### PR TITLE
Add integration tests

### PR DESCRIPTION
**Pull request by danielstjules**
Quite a few tests (acceptance, regression, etc) require either manual invocation or visual confirmation after running the test suite to confirm their success or failure. This PR updates some of those to use a helper that invokes mocha and verifies its output. Parts were extracted/modified from https://github.com/mochajs/mocha/pull/1618

Edit: For clarification. This mostly deals with test files that previously were never ran (had to be manually done), or which required close watch of the test results and console output, as their failure wouldn't trigger a failing build.

### CODE REVIEW

File: test/integration/fixtures/options/bail.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+describe('suite2', function() {
+  // TODO: When uncommented, the hook below is ran and throws an exception
+  // despite the previous failure with the bail flag. This is a bug. Uncomment
+  // once resolved
+
+  // before(function(done) {
+  //   throw new Error('this hook should not be displayed');
+  // });
+
```
Comment (danielstjules) [inline test/integration/fixtures/options/bail.js]: Found a bug here. Will open a new issue for this.

**Comment by danielstjules** [comment]
@mochajs/mocha :) Should help with CI.

**Comment by a8m** [comment]
Wow, this is huge(and looks pretty good actually).
I'll look it deeply this weekend. (I think I'll have coffee next to me, a big one. haha :smile:)

**Comment by dasilvacontin** [comment]
And popcorn. :tada:

**Comment by danielstjules** [comment]
ping :)

File: test/integration/helpers.js [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  /**
+   * Invokes the mocha binary for the given fixture using the JSON reporter,
+   * returning the parsed output, as well as exit code.
+   *
+   * @param {string}   fixturePath
+   * @param {array}    args
+   * @param {function} fn
+   */
+  runMochaJSON: function(fixturePath, args, fn) {
+    var path, args, output, mocha, listener;
```
Comment (a8m) [inline test/integration/helpers.js]: `output, mocha, listener` are unused

File: test/integration/helpers.js [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+   *   code:    1,
+   *   output:  '...'
+   * }
+   *
+   * @param {string}   fixturePath
+   * @param {array}    args
+   * @param {function} fn
+   */
+  runMocha: function(fixturePath, args, fn) {
+    var path, args, output, mocha, listener;
```
Comment (a8m) [inline test/integration/helpers.js]: same

**Comment by a8m** [comment]
LGTM!
Great work @danielstjules. I really like it.
Thx.

File: test/integration/fixtures/timeout.js [hunk 4]
```diff
@@ -1,16 +1,16 @@
...
-  it('should be honored with async suites', function(done){
+  it('should be honored with async suites', function(done) {
     sleep(2);
     done();
   });
 
-  function sleep(ms){
+  function sleep(ms) {
     var start = Date.now();
     while(start + ms > Date.now())continue;
```
Comment (a8m) [inline test/integration/fixtures/timeout.js]: This `continue` is unnecessary here(because this loop doesn't have a body)

**Comment by danielstjules** [comment]
Done! I was thinking I'd merge this evening unless someone else has some recommendations.

cc @mochajs/owners

**Pull request merged**